core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -624,6 +625,26 @@ public synchronized void updateMetaData(final IndexMetaData metadata) {
                 }
             }
             if (refreshTask.getInterval().equals(indexSettings.getRefreshInterval()) == false) {
+                // once we change the refresh interval we schedule yet another refresh
+                // to ensure we are in a clean and predictable state.
+                // it doesn't matter if we move from or to <code>-1</code>  in both cases we want
+                // docs to become visible immediately
```

Comment:
yeah so I wonder if we should rather build it into the refresh listener that is forces a refresh if refreshes are disabled

Reply:
I'm not sure people setting `wait_for` will want a `immediately` executed because refreshes are set to -1. I'm +1 to solving it in the refresh listeners - out of scope for this PR of course.